### Facebook Login Controller

<!-- > Visit https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/

> " Using unknown in Catch Variables Users running with the --strict flag may see new errors around catch variables being unknown, especially if the existing code assumes only Error values have been caught. This often results in error messages such as:

> Property 'message' does not exist on type 'unknown'. Property 'name' does not exist on type 'unknown'. Property 'stack' does not exist on type 'unknown'.

> To get around this, you can specifically add runtime checks to ensure that the thrown type matches your expected type. Otherwise, you can just use a type assertion,

> add an explicit : any to your catch variable, or turn off --useUnknownInCatchVariables."

> you can simply add in tsconfig.json:

> `"compilerOptions": { "useUnknownInCatchVariables": false }`

> or you can use like:

> `catch (e: any) {console.log(e.message)}` -->

- Para comecar, vamos criar nosso teste sem nenhum tipo de refactor e deixando o codigo "apenas" para passar nos testes.
- Nesse primeiro momento queremos apenas que os comportamentos sejam esperados e nao iremos implementar "boas praticas" de codigo, pois nesta secao utilizaremos para refatorar, diversos design patterns que sao de altissimo nivel de qualidade.

### Setup:

> WARN: Lembrando que utilizamos o **mockResolvedValue** ou **mockReturnValue** quando queremos mockar o caso de sucesso daquela funcao.

```
  let sut: FacebookLoginController
  let facebookAuth: MockProxy<FacebookAuthentication>

  beforeAll(() => {
    facebookAuth = mock()
    facebookAuth.perform.mockResolvedValue(new AccessToken('any_value'))
  })

  beforeEach(() => {
    sut = new FacebookLoginController(facebookAuth)
  })
```

### Helpers

> o Type HttpResponse e um helper para dizermos que nosso controlador sempre deve retornar uma resposta para o client solicitante, normalmente quando utilizado Express sem desacoplamento vemos o Controller retornando o seguinte: `Promise<Response>`, onde o **Response** e do tipo proprio Express.

> A classe ServerError nao e nada mais que uma classe para lidar com Erros do Servidor e retornar uma mensagem mais amigavel para o client solicitante ao inves do stacktrace.
> Um exemplo de fluxo seria caso o metodo do servico **facebookAuth.perform** tentasse acessar a API do Facebook e por algum motivo aparecer um erro (manutencao), nos **DEFINIMOS** que retornaremos um **AuthenticationError()**, porem se ocorrer um erro de infra como ficar sem internet, aquela promise nao vai executar e vai "estourar" um erro no console com o stacktrace, com isso temos nosso erro de Servidor.

```
type HttpResponse = {
  statusCode: number
  data: any
}

class ServerError extends Error {
  constructor (error?: Error) {
    super('Server failed. Try again soon')
    this.name = 'ServerError'
    this.stack = error?.stack
  }
}
```

### Testes
> Estamos testando diversos tipos de erros de StatusCode por parte do cliente, caso de sucesso e erro de infraestrutura (500).
> Pode-se notar que nao temos um setup muito diferenciado ou ate mesmo diversos mocks nos testes, porque nossa classe de producao esta bem "feia" mas cumprindo os testes. (IREMOS REFATORAR.)

```
it('should return 400 if token is empty', async () => {
    const httpResponse = await sut.handle({ token: '' })

    expect(httpResponse).toEqual({
      statusCode: 400,
      data: new Error('The field token is required')
    })
  })

  it('should return 400 if token is null', async () => {
    const httpResponse = await sut.handle({ token: null })

    expect(httpResponse).toEqual({
      statusCode: 400,
      data: new Error('The field token is required')
    })
  })

  it('should return 400 if token is undefined', async () => {
    const httpResponse = await sut.handle({ token: undefined })

    expect(httpResponse).toEqual({
      statusCode: 400,
      data: new Error('The field token is required')
    })
  })

  it('should call FacebookAuthentication with correct params', async () => {
    await sut.handle({ token: 'any_token' })

    expect(facebookAuth.perform).toHaveBeenCalledWith({ token: 'any_token' })
    expect(facebookAuth.perform).toHaveBeenCalledTimes(1)
  })

  it('should return 401 if authentication fails', async () => {
    facebookAuth.perform.mockResolvedValueOnce(new AuthenticationError())
    const httpResponse = await sut.handle({ token: 'any_token' })

    expect(httpResponse).toEqual({
      statusCode: 401,
      data: new AuthenticationError()
    })
  })

  it('should return 200 if authentication succeeds', async () => {
    const httpResponse = await sut.handle({ token: 'any_token' })

    expect(httpResponse).toEqual({
      statusCode: 200,
      data: {
        accessToken: 'any_value'
      }
    })
  })

  it('should return 500 if authentication throws', async () => {
    const error = new Error('infra_error')
    facebookAuth.perform.mockRejectedValueOnce(error)
    const httpResponse = await sut.handle({ token: 'any_token' })

    expect(httpResponse).toEqual({
      statusCode: 500,
      data: new ServerError(error)
    })
  })
```


### Classe de Producao

```
class FacebookLoginController {
  constructor (private readonly facebookAuthentication: FacebookAuthentication) {}

  async handle (httpRequest: any): Promise<HttpResponse> {
    try {
      if (httpRequest.token === '' || httpRequest.token === null || httpRequest.token === undefined) {
        return {
          statusCode: 400,
          data: new Error('The field token is required')
        }
      }
      const result = await this.facebookAuthentication.perform({ token: httpRequest.token })
      if (result instanceof AccessToken) {
        return {
          statusCode: 200,
          data: {
            accessToken: result.value
          }
        }
      } else {
        return {
          statusCode: 401,
          data: result
        }
      }
    } catch (error) {
      return {
        statusCode: 500,
        data: new ServerError(error)
      }
    }
  }
}
```
